externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpBase.td

```diff
@@ -0,0 +1,30 @@
+#ifndef TORCH_MLIR_DIALECT_TCP_BASE
+#define TORCH_MLIR_DIALECT_TCP_BASE
+
+include "mlir/IR/OpBase.td"
+
+def Tcp_Dialect : Dialect {
+  let name = "tcp";
+  let cppNamespace = "::mlir::torch::tcp";
+  let description = [{
+    Tensor Compute Primitives (TCP) dialect.
+
+    TCP is a mid-level dialect in MLIR that enables transformations that
+    complement those in existing dialects, by supporting ops at a
+    high-level of abstraction.
+  }];
+
+  let dependentDialects = [];
+}
+
+def Tcp_Scalar : AnyTypeOf<[AnyFloat, AnySignlessIntegerOrIndex, AnyComplex]>;
```

Comment:
This will likely become clear as we design more ops, but IMO we should keep the `index` type distinct from other non-index types.

Reply:
+1. The one exception where `tensor<index>` should exist is for "gather/scatter indices". We definitely don't want to ever use `tensor<?xindex>` as a "shape" which happens in MHLO and is the source of much brittleness.

Reply:
Good points regarding `index`. However, in addition to gather / scatter, we should probably allow some binary operations on index types, for example, max, min. Wdyt?

Updated this PR to not include `index` types.

Reply:
> we should probably allow some binary operations on index types, for example, max, min

Yes, but I'm not convinced that we need these operations on tensors containing index types.

Reply:
There was a comment in one of our docs that seemed to be talking about a use-case for this. I might be mistaken though. 

Anyways, I have removed it in this PR. We can table this discussion until we have a real use-case with index type beyond Gather / Scatter.